File: AutoMix Software/AutoMix Data Management/AudioDataExtraction.cpp
﻿// Copyright (C) 2016-2017 LesProjecteurs - All Rights Reserved
// Maxime STEVENOT, Guillaume HANNES, Jordan ERNULT, Louis CARLIER, Pierre GABON
// 
// This file is part of AutoMix Software.
// 
// Unauthorized copying of this file, via any medium is strictly prohibited.
// You should have received a copy of the License along with this program.

#include "stdafx.h"
#include "AudioDataExtraction.h"

#undef GetTempPath

using namespace System::Diagnostics;
using namespace System::IO;
using namespace System;
using namespace Concurrency;

namespace AutoMixDataManagement {

	AudioDataExtraction::AudioDataExtraction()
	{
		String^ extractorpath = System::IO::Directory::GetCurrentDirectory() + "\\essentia_streaming_extractor_music.exe";
		String^ temppath = Path::GetTempPath() + "AutomixSoftware";
		if (!Directory::Exists(temppath)) {
			_tempDirectory = Directory::CreateDirectory(temppath);
		}
		_startInfo = gcnew ProcessStartInfo(extractorpath);
		_startInfo->UseShellExecute = false;
		_startInfo->WorkingDirectory = _tempDirectory->FullName;
		_startInfo->WindowStyle = ProcessWindowStyle::Hidden;

		String^ profileName = _tempDirectory->FullName + "\\profile.yaml";
		StreamWriter^ sw = gcnew StreamWriter(profileName);
		sw->Write("outputFormat: json\noutputFrames: 0\nlowlevel:\n    stats: [ \"mean\" ]\n    mfccStats: [\"mean\"]\n\
			gfccStats : [\"mean\"]\nrhythm :\n    stats : [\"mean\", \"var\", \"median\", \"min\", \"max\"]\ntonal :\n\
			stats : [\"mean\", \"var\", \"median\", \"min\", \"max\"]");
		sw->Close();
	}

	AudioDataExtraction::~AudioDataExtraction()
	{
		_tempDirectory->Delete();
	}

	void AudioDataExtraction::extractData(Track^ track)
	{
		String^ parameters = track->Path + " " + track->Name + ".json " + _tempDirectory->FullName + "\\profile.yaml ";
		_startInfo->Arguments = parameters;
		Process^ extractor = gcnew Process;
		extractor->StartInfo = _startInfo;
		extractor->Start();
		extractor->WaitForExit();
	}

	void AudioDataExtraction::extractData(TrackCollection^ trackCollection)
	{
		TrackCollection::Enumerator i = trackCollection->GetEnumerator();
		Track^ last = gcnew Track();
		while (i.MoveNext())
		{
			last = i.Current;
		}
		i.Reset();
		i.MoveNext();
		parallel_for_each(i.Current, last, [&](Track^ track) {
			extractData(track);
			i.MoveNext();
		});
	}
}

File: Automix Data Management/Extraction/AudioDataExtraction.cs
﻿// Copyright (C) 2016-2017 LesProjecteurs - All Rights Reserved
// Maxime STEVENOT, Guillaume HANNES, Jordan ERNULT, Louis CARLIER, Pierre GABON
// 
// This file is part of Automix Software.
// 
// Unauthorized copying of this file, via any medium is strictly prohibited.
// You should have received a copy of the License along with this program.

using System;
using System.Diagnostics;
using System.IO;
using System.ComponentModel;
using System.Threading;
using System.Threading.Tasks;
using Automix_Data_Management.Model;
using Automix_Data_Management.Storage;

namespace Automix_Data_Management.Extraction
{
    internal class AudioDataExtraction : IAudioDataExtraction
    {
        private DirectoryInfo _tempDirectory;
        public static int ExploredTracks;

        public AudioDataExtraction()
        {
            new DataBase();
            InitExecConfiguration();
            ExploredTracks = 0;
        }

        public void ExtractData(BackgroundWorker backgroundWorker, TrackCollection trackCollection)
        {
            var cancellationTokenSource = new CancellationTokenSource();
            var options = new ParallelOptions
            {
                CancellationToken = cancellationTokenSource.Token,
                MaxDegreeOfParallelism = (int)Math.Ceiling(Environment.ProcessorCount / 2.0)
            };
            ExploredTracks = 0;

            try
            {
                Parallel.ForEach(trackCollection, options, (track) =>
                {
                    if (cancellationTokenSource.IsCancellationRequested)
                    {
                        return;
                    }

                    IExtraction extraction;
                    var dataBase = new DataBase();

                    if (dataBase.IsInDataBase(track))
                    {
                        extraction = new DataBaseExtraction(dataBase);
                    }
                    else
                    {
                        extraction = new ExecutableExtraction(_tempDirectory);
                    }
                    if (!backgroundWorker.CancellationPending)
                    {
                        extraction.ExtractData(backgroundWorker, trackCollection.Count, cancellationTokenSource, track);
                    }
                    else
                    {
                        cancellationTokenSource.Cancel();
                    }
                });
            }
            catch (OperationCanceledException)
            {
                //TODO log file
            }
        }

        private void InitExecConfiguration()
        {
            var tempPath = Path.GetTempPath() + "AutomixSoftware";

            if (!Directory.Exists(tempPath))
            {
                _tempDirectory = Directory.CreateDirectory(tempPath);
            }
            else
            {
                Directory.Delete(tempPath, true);
                _tempDirectory = Directory.CreateDirectory(tempPath);
            }

            try
            {
                var profileName = _tempDirectory.FullName + "\\profile.yaml";
                var sw = new StreamWriter(profileName);
                sw.Write("outputFormat: json\noutputFrames: 0\nlowlevel:\n    stats: [ \"mean\" ]\n    mfccStats: [\"mean\"]"
                    + "\n    gfccStats: [\"mean\"]\nrhythm :\n    stats : [\"mean\"]\ntonal :\n    stats: [\"mean\"]");
                sw.Close();
            }
            catch (IOException e)
            {
                Debug.WriteLine(e.Message);
            }
        }
    }
}